nixos/modules/services/networking/prosody.nix

```diff
@@ -1,887 +1,675 @@
-    };
+  acmeHosts = unique (mapAttrsToList (domain: hostOpts: hostOpts.useACMEHost) (filterAttrs (_: v: v.useACMEHost != null) cfg.virtualHosts));
+in
+{
+  imports = [
+    (mkRenamedOptionModule [ "services" "prosody" "allowRegistration" ] [ "services" "prosody" "settings" "allow_registration" ])
+    (mkRenamedOptionModule [ "services" "prosody" "httpPorts" ] [ "services" "prosody" "settings" "http_ports" ])
+    (mkRenamedOptionModule [ "services" "prosody" "httpInterfaces" ] [ "services" "prosody" "settings" "http_interfaces" ])
+    (mkRenamedOptionModule [ "services" "prosody" "httpsPorts" ] [ "services" "prosody" "settings" "https_ports" ])
+    (mkRenamedOptionModule [ "services" "prosody" "httpsInterfaces" ] [ "services" "prosody" "settings" "https_interfaces" ])
+    (mkRenamedOptionModule [ "services" "prosody" "c2sRequireEncryption" ] [ "services" "prosody" "settings" "c2s_require_encryption" ])
+    (mkRenamedOptionModule [ "services" "prosody" "s2sRequireEncryption" ] [ "services" "prosody" "settings" "s2s_require_encryption" ])
+    (mkRenamedOptionModule [ "services" "prosody" "s2sSecureAuth" ] [ "services" "prosody" "settings" "s2s_secure_auth" ])
+    (mkRenamedOptionModule [ "services" "prosody" "s2sInsecureDomains" ] [ "services" "prosody" "settings" "s2s_insecure_domains" ])
+    (mkRenamedOptionModule [ "services" "prosody" "s2sSecureDomains" ] [ "services" "prosody" "settings" "s2s_secure_domains" ])
+    (mkRenamedOptionModule [ "services" "prosody" "extraModules" ] [ "services" "prosody" "settings" "modules_enabled" ])
+    (mkRenamedOptionModule [ "services" "prosody" "extraPluginPaths" ] [ "services" "prosody" "settings" "plugin_paths" ])
+    (mkRenamedOptionModule [ "services" "prosody" "ssl" "key" ] [ "services" "prosody" "settings" "ssl" "key" ])
+    (mkRenamedOptionModule [ "services" "prosody" "ssl" "cert" ] [ "services" "prosody" "settings" "ssl" "certificate" ])
+    (mkRenamedOptionModule [ "services" "prosody" "ssl" "extraOptions" ] [ "services" "prosody" "settings" "ssl" ])
+    (mkRenamedOptionModule [ "services" "prosody" "admins" ] [ "services" "prosody" "settings" "admins" ])
+    (mkRenamedOptionModule [ "services" "prosody" "authentication" ] [ "services" "prosody" "settings" "authentication" ])
+    (mkChangedOptionModule [ "services" "prosody" "disco_items" ] [ "services" "prosody" "settings" "disco_items" ] (config:
+      map ({ url, description }: [ url description ]) config.services.prosody.disco_items
+    ))
+    (mkChangedOptionModule [ "services" "prosody" "uploadHttp" ] [ "services" "prosody" "components" ] (config:
+      let
+        cfg = config.services.prosody;
+      in
+      {
+        ${cfg.uploadHttp.domain} = {
+          module = "http_upload";
+          # these values are to preserve compatibility with this module pre nixos 23.11
+          settings = {
+            http_upload_file_size_limit = cfg.uploadHttp.uploadFileSizeLimit or "50 * 1024 * 1024";
```

Comment:
Not sure if it matters but setting these as string make them quoted in config:
```
-    http_upload_file_size_limit = 10485760
-    http_upload_expire_after = 60 * 60 * 24 * 7
+http_upload_expire_after = "60 * 60 * 24 * 7"
+http_upload_file_size_limit = "10485760"
```
We might need to make these integers somehow...? need to check how prosody handles that.

(Not using default, this is my config:)
```
    uploadHttp = {
      domain = "tmp.xxx";
      # Prosody's HTTP parser limit on body size
      uploadFileSizeLimit = "10485760";
      userQuota = 100 * 1024 * 1024;
      uploadExpireAfter = "60 * 60 * 24 * 7";
      httpUploadPath = "/var/lib/prosody/upload";
    };
```

Reply:
I must have been tired when writing that code... I remember thinking "why aren't these integers in lua?" 😆

Good catch, thanks!

Reply:
This one might require an explicit cast in the mkChangedOptionModule as the old setting was a string; it's not just defaults...
Unfortunately `lib.toInt "2 * 3"` fails because 2 * 3 isn't a plain int, so I'm not quite sure how to fix that :/

(also can't seem t be able to resolve/unresolve these conversations, so hope you see this :grin:)

Reply:
Oh yeah I see now.... ugh... ok I have a dirty idea to handle this. Thanks.

Reply:
ok i pushed a hacky fix for this so the given config:
```
uploadHttp = {
  domain = "upload.example.org";
  # fallback to the default values which are strings representing math and ints...
};
```
produces the following output file:
```
Component "upload.example.org" "http_upload"
http_upload_expire_after = 60 * 60 * 24 * 7
http_upload_file_size_limit = 50 * 1024 * 1024
http_upload_path = "/var/lib/prosody"
```

i left this as a separate commit for now so it is easier to see what is going on

we should really drop support for this compat stuff in a release or two... it would be a shame to let this compat code sit around for the next 5 years :sweat:

Reply:
Thanks!
The fix makes sense to me (I'd nitpick and rename __compat to __compat_int or something to hint at why we do this in the code, but it's probably fine)

And.. yeah, probably want to get rid of this next year; there's a huge wall of warnings when building one's config through these remaps so hopefully people will be encouraged to migrate faster than later (I sure will!)